### ISSUE TITLE
twisted.test.test_defer.DeferredTestCase.testMaybeDeferred should be faster and four different tests

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2894]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2894|
|Type|enhancement|
|Created|2007-11-13 13:37:38Z|
|Branch|https://github.com/twisted/twisted/tree/test-maybedeferred-2894|

```testMaybeDeferred``` uses ```reactor.callLater(0.2``` twice, making it take at least 0.4 seconds to run.  This could easily be shortened so that it succeeds synchronously (it's trying to test the uncalled ```Deferred``` case for success and error, but it could do this by calling back the ```Deferred``` after ```maybeDeferred``` has gotten a hold of it).

It's also testing four different conditions, so it should be broken into separate test methods.

<details><summary>Searchable metadata</summary>

```
trac-id__2894 2894
type__enhancement enhancement
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__branches_test_maybedeferred_2894 branches/test-maybedeferred-2894
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1194961058000000 1194961058000000
changetime__1198922114000000 1198922114000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_1"></a>|@therve commented|
|-|-|

(In [22091]) Branching to 'test-maybedeferred-2894'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve commented|
|-|-|

(In [22092]) Fix testMaybeDeferred: split it in 4, add some docstrings.

Refs [#2894](https://github.com/twisted/twisted/issues/2894)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve removed owner|
|-|-|

This is ready to review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @therve<br>@exarkun set status to `new`|
|-|-|

Looks good

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set status to `closed`|
|-|-|

(In [22179]) Merge test-maybedeferred-2894

Author: therve
Reviewer: exarkun
Fixes [#2894](https://github.com/twisted/twisted/issues/2894)

Split the tests of maybeDeferred into 4 smaller tests, and remove the callLater
from the tests.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
